### PR TITLE
[sumami-holders] Add sUmami holders strategy

### PR DESCRIPTION
**Pull request by arugulo**
Fixes # .

Changes proposed in this pull request:
- 
- 
-

### CODE REVIEW

File: src/strategies/sumami-holders/examples.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
+[
+  {
+    "name": "Example query",
+    "strategy": {
+      "name": "erc20-balance-of",
```
Comment (ChaituVR) [inline src/strategies/sumami-holders/examples.json:5]:
```suggestion
      "name": "sumami-holders",
```

File: src/strategies/sumami-holders/examples.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      "params": {
+        "address": "0x6b175474e89094c44da98b954eedeac495271d0f",
+        "symbol": "DAI",
+        "decimals": 18
```
Comment (ChaituVR) [inline src/strategies/sumami-holders/examples.json:9]:
```suggestion
        "address": "0xe6d557d416ff5640235119369c7e26AA18a906D7",
        "marinateLevels": [0, 1, 2, 3],
        "symbol": "sUMAMI",
        "marinateAddress": "0x190a6b6E8e4D9B8324E1F97127c588C5b082d94b",
        "decimals": 9
```

File: src/strategies/sumami-holders/index.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    inputs: [
+      {internalType: "address", name: "account", type: "address"},
+      {internalType: "uint32", name: "level", type: "uint32"}
+    ],
+    name: "stakedBalance",
+    outputs: [{internalType: "uint256",name: "",type: "uint256"}],
+    stateMutability: "view",
+    type: "function"
+  }
+];
```
Comment (ChaituVR) [inline src/strategies/sumami-holders/index.ts:27]:
```suggestion
const abi = [
  'function balanceOf(address account) view returns (uint256)',
  'function stakedBalance(address account, uint32 level) view returns (uint256)'
];
```

File: src/strategies/sumami-holders/index.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
+import { formatUnits } from '@ethersproject/units';
+import { multicall } from '../../utils';
+import { strategy as erc20BalanceOfStrategy } from '../erc20-balance-of';
+
+export const author = 'arugulo';
+export const version = '0.1.1';
```
Comment (ChaituVR) [inline src/strategies/sumami-holders/index.ts:6]:
```suggestion
export const version = '0.1.0';
```

File: src/strategies/sumami-holders/index.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        { blockTag }
+      )
+    )
+  );
+  
+  return Object.fromEntries(
+    Object.entries(sUmamiBalances).map((address, index) => [
+      address[0],
+      address[1] + 
+        marinateBalances.reduce((prev: number, cur: any, idx: number) =>
```
Comment (ChaituVR) [inline src/strategies/sumami-holders/index.ts:69]:
```JS
src/strategies/sumami-holders/index.ts#L70

'idx' is defined but never used
```

File: src/strategies/sumami-holders/index.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    space,
+    network,
+    provider,
+    addresses,
+    options,
+    snapshot
+  );
+
+  const marinateBalances = await Promise.all(
+    options.marinateLevels.map((level: number) =>
```
Comment (ChaituVR) [inline src/strategies/sumami-holders/index.ts:49]: We should have a limit of 4 marinateLevels here else some other space may pass in some big amount of requests into this
There is a limit of 5 calls
https://github.com/snapshot-labs/snapshot-strategies#checklist-for-adding-a-new-strategy

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/strategies/sumami-holders/index.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  space,
+  network,
+  provider,
+  addresses,
+  options,
+  snapshot
+) {
+  const blockTag = typeof snapshot === 'number' ? snapshot : 'latest';
+  if (options.marinateLevels.length > 4) {
+    return;
```
Comment (ChaituVR) [inline src/strategies/sumami-holders/index.ts:24]:
```suggestion
    return [];
```

File: src/strategies/sumami-holders/README.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+```JSON
+{
+    "address": "0xe6d557d416ff5640235119369c7e26AA18a906D7",
+    "marinateLevels": [0, 1, 2, 3],
+    "symbol": "sUMAMI",
+    "marinateAddress": "0x190a6b6E8e4D9B8324E1F97127c588C5b082d94b",
+    "decimals": 9
+}
+```
```
Comment (ChaituVR) [inline src/strategies/sumami-holders/README.md:17]:
```suggestion
```
Note: if `marinateLevels` length is more than 4, the strategy will return 0 voting power
```

File: src/strategies/sumami-holders/index.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  }
+  
+  const sUmamiBalances = await erc20BalanceOfStrategy(
```
Comment (ChaituVR) [inline src/strategies/sumami-holders/index.ts:27]: The strategy fails with the following error:
```
call revert exception (method="balanceOf(address)", errorArgs=null, errorName=null, errorSignature=null, reason=null, code=CALL_EXCEPTION, version=abi/5.4.0)

      at Logger.Object.<anonymous>.Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:213:28)

```
Looks like addresses in examples.json are not valid tokens/contracts on mainnet?

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/strategies/sumami-holders/index.ts (hunk 9 above)
Comment (arugulo) [reply src/strategies/sumami-holders/index.ts:27]: Right it's on arbitrum. I've changed the network in the example

**Comment by ChaituVR** [comment]
@arugulo It works in your local?? im getting this error:
```
missing revert data in call exception (error={"reason":"processing response error","code":"SERVER_ERROR","body":"{\"jsonrpc\": \"2.0\", \"id\": 44, \"error\": {\"code\": -32000, \"message\": \"This request is not supported because your node is running with state pruning. Run with --pruning=archive.\"}}","error":{"code":-32000},
```

**Comment by arugulo** [comment]
Sorry forgot to push updated example. It should pass tests now

**Review by bonustrack — APPROVED** [review]

**Review by ChaituVR — APPROVED** [review]

**Pull request merged**